src/Razor/src/Microsoft.NET.Sdk.Razor/ReferenceResolver.cs

```diff
@@ -15,155 +17,138 @@ namespace Microsoft.AspNetCore.Razor.Tasks
 
             return applicationParts;
         }
 
-        private DependencyClassification Resolve(ClassifiedAssemblyItem classifiedItem)
+        private Classification Resolve(AssemblyItem assemblyItem)
         {
-            if (classifiedItem.DependencyClassification != DependencyClassification.Unknown)
+            if (_classifications.TryGetValue(assemblyItem, out var classification))
             {
-                return classifiedItem.DependencyClassification;
+                return classification;
             }
 
-            if (classifiedItem.AssemblyItem == null)
+            // Initialize the dictionary with a value to short-circuit recursive references.
+            classification = Classification.Unknown;
+            _classifications[assemblyItem] = classification;
+
+            if (assemblyItem.Path == null)
             {
-                // We encountered a dependency that isn't part of this assembly's dependency set. We'll see if it happens to be an MVC assembly.
-                // This might be useful in scenarios where the app does not have a framework reference at the entry point,
-                // but the transitive dependency does.
-                classifiedItem.DependencyClassification = _mvcAssemblies.Contains(classifiedItem.Name) ?
-                    DependencyClassification.MvcReference :
-                    DependencyClassification.DoesNotReferenceMvc;
-
-                return classifiedItem.DependencyClassification;
+                // We encountered a dependency that isn't part of this assembly's dependency set. We'll see if it happens to be an MVC assembly
+                // since that's the only useful determination we can make given the assembly name.
+                classification = _mvcAssemblies.Contains(assemblyItem.AssemblyName) ?
+                    Classification.IsMvc :
+                    Classification.DoesNotReferenceMvc;
             }
-
-            if (classifiedItem.AssemblyItem.IsSystemReference)
+            else if (assemblyItem.IsSystemReference)
             {
                 // We do not allow transitive references to MVC via a framework reference to count.
                 // e.g. depending on Microsoft.AspNetCore.SomeThingNewThatDependsOnMvc would not result in an assembly being treated as
                 // referencing MVC.
-                classifiedItem.DependencyClassification = _mvcAssemblies.Contains(classifiedItem.Name) ?
-                    DependencyClassification.MvcReference :
-                    DependencyClassification.DoesNotReferenceMvc;
-
-                return classifiedItem.DependencyClassification;
+                classification = _mvcAssemblies.Contains(assemblyItem.AssemblyName) ?
+                    Classification.IsMvc :
+                    Classification.DoesNotReferenceMvc;
             }
-
-            if (_mvcAssemblies.Contains(classifiedItem.Name))
+            else if (_mvcAssemblies.Contains(assemblyItem.AssemblyName))
```

Comment:
Wont everything MVC be eaten by the above `else if`?

Reply:
We have assemblies like Newtonsoft.Json that aren't part of the fx.